### PR TITLE
Add `vertical-front-matter` feature

### LINKED ISSUE #3325: Markdown Preview: transpose Frontmatter Table
### Summary

The Markdown preview renders the yaml frontmatter in a table, but the options as header and below the values. but when there are many frontmatter values it not looks very nice. in the screenshot the exceprt is not visible.

### suggested Improvement

it would be better when for each value there is a row.

### Link

https://github.com/gatsbyjs/gatsby/blob/2a00a474d15b320f24802210e8abde52bc44f50c/docs/blog/2020-07-07-wordpress-source-beta/index.mdx

#### Current State

Bildschirmfoto 2020-07-08 um 13 59 17

#### With the feature

Bildschirmfoto 2020-07-08 um 14 20 38

### PR DESCRIPTION
**Pull request by kidonng**
Close #3325 
Test: https://github.com/gatsbyjs/gatsby/blob/master/docs/blog/2020-07-07-wordpress-source-beta/index.mdx

Screenshot: 
image

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -164,6 +164,7 @@ Thanks for contributing! 🦋🙌
 - [](# "esc-to-deselect-line") [Adds a keyboard shortcut to deselect the current line: <kbd>esc</kbd>.](https://github.com/sindresorhus/refined-github/issues/1590)
 - [](# "html-preview-link") [Adds a link to preview HTML files.](https://user-images.githubusercontent.com/44045911/67634792-48995980-f8fb-11e9-8b6a-7b57d5b12a2f.png)
 - [](# "action-used-by-link") [Lets you see how others are using the current Action in the Marketplace.](https://user-images.githubusercontent.com/8360597/80250140-86d9c080-8673-11ea-9d28-f62faf9fd3d4.png)
+- [](# "vertical-front-matter") [Show Markdown front matter as vertical table.](https://user-images.githubusercontent.com/184316/86918308-c9655200-c126-11ea-8e10-31013667dd76.png)
```
Comment (muescha) [inline readme.md:167]: the screenshots (here and in `vertical-front-matter.tsx` ) are the screenshot from my issue, but your screenshot looks different

(no header, and left `th` are not have a `text-align: right`)

Comment (kidonng) [reply readme.md:167]: You are right, I wrote the description first and forgot to update the screenshot here.

File: source/features/vertical-front-matter.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	while (select.exists('th', thead)) {
+		const th = select('th', thead)!;
+		th.classList.add('rgh-front-matter-name');
+		tbody.append(
+			<tr>
+				{th}
+				{select('td', tbody)!}
+			</tr>
+		);
+	}
```
Comment (fregante) [inline source/features/vertical-front-matter.tsx:26]: Can you find a proper transpose algorithm? There’s a lot of `select`ion here

Comment (kidonng) [reply source/features/vertical-front-matter.tsx:26]: > There’s a lot of `select`ion here.

Can I just use `tr.children[0]`? So there is no selection.

Comment (muescha) [reply source/features/vertical-front-matter.tsx:36]: `thead.children[index]` `tbody.children[index]` 

or use [`select.all`](https://github.com/fregante/select-dom#selectallselector-baseelements--document) to get 2 arrays back:

```javascript
const thead = select.all('thead tr th', table)!;
const tbody = select.all('tbody', table)!;
```

and then `thead[index]`, `tbody[index]`

**Comment by muescha** [comment]
since the comments are marked as resolved:

would it better to use [`select.all`](https://github.com/fregante/select-dom#selectallselector-baseelements--document) to get 2 arrays back:

```javascript
const thead = select.all('thead tr th', table)!;
const tbody = select.all('tbody', table)!;
```

and then `thead[index]`, `tbody[index]`

**Comment by muescha** [comment]
you can also add only one time a css style `rgh-vertical-front-matter-table` to the table tag and change the css to, then you not need to alter each `th` in the loop:

```css
.rgh-vertical-front-matter-table tbody tr th {
	text-align: right;
}
```

**Comment by muescha** [comment]
and code simplifies to:

```javascript
const thead = select.all('thead tr th', table)!;
const tbody = select.all('tbody', table)!;

for (let i = 0; i < count; i++) {
		tbody.append(
			<tr>
				{thead[index]}
				{tbody[index]}
			</tr>
		);
}
```

File: source/features/vertical-front-matter.css [hunk 3]
```diff
@@ -1,3 +1,3 @@
-.rgh-front-matter-name {
+.rgh-vertical-front-matter-table th {
```
Comment (muescha) [inline source/features/vertical-front-matter.css:1]: i would use the `tbody` if there later is a header, this gets confused
```suggestion
.rgh-vertical-front-matter-table tbody tr th {
```

**Comment by kidonng** [comment]
> would it better to use [`select.all`](https://github.com/fregante/select-dom#selectallselector-baseelements--document) to get 2 arrays back:

This matched nested tables. 😕

**Comment by muescha** [comment]
in css this would work:

```css
select.all("table[data-table-type="yaml-metadata"] > thead > tr > th")
select.all("table[data-table-type="yaml-metadata"] > tbody > tr > td")
```

**Comment by kidonng** [comment]
Just made some perf test: https://jsperf.com/qsa-vs-em

Correct me if I'm wrong, but I don't see the benefit.

**Comment by fregante** [comment]
It's not just about performance but also about readability. 

This looks good and short if it works:

https://github.com/sindresorhus/refined-github/pull/3327#issuecomment-655599475

**Comment by kidonng** [comment]
```ts
	// Table > thead
	const thead = table.firstElementChild!;
	// Table > thead > tr > th
	const theadCells = select.all('th', thead);
	if (theadCells.length <= 4) {
		return;
	}

	// Table > tbody
	const tbody = table.lastElementChild!;
	// Table > tbody > tr
	const tbodyRow = tbody.firstElementChild!;
	const tbodyCells = select.all('table[data-table-type="yaml-metadata"] > tbody > tr > td' ,tbodyRow);
	for (let i = 0; i < theadCells.length; i++) {
		tbody.append(
			<tr>
				{theadCells[i]}
				{tbodyCells[i]}
			</tr>
		);
	}
```

I don't think they really differ in readability.

**Comment by muescha** [comment]
```diff
-	// Table > thead
-	const thead = table.firstElementChild!;
-	// Table > thead > tr > th
-	const theadCells = select.all('th', thead);
+	const theadCells = select.all('table[data-table-type="yaml-metadata"] > tbody > tr > td');
	if (theadCells.length <= 4) {
		return;
	}

-	// Table > tbody
-	const tbody = table.lastElementChild!;
-	// Table > tbody > tr
-	const tbodyRow = tbody.firstElementChild!;
-	const tbodyCells = select.all('table[data-table-type="yaml-metadata"] > tbody > tr > td' ,tbodyRow);
+	const tbodyCells = select.all('table[data-table-type="yaml-metadata"] > tbody > tr > td');
	for (let i = 0; i < theadCells.length; i++) {
		tbody.append(
			<tr>
				{theadCells[i]}
				{tbodyCells[i]}
			</tr>
		);
	}
```

```javascript
	const theadCells = select.all('table[data-table-type="yaml-metadata"] > tbody > tr > td');
	if (theadCells.length <= 4) {
		return;
	}

	const tbodyCells = select.all('table[data-table-type="yaml-metadata"] > tbody > tr > td');
	for (let i = 0; i < theadCells.length; i++) {
		tbody.append(    // <-- this needs to changed
			<tr>
				{theadCells[i]}
				{tbodyCells[i]}
			</tr>
		);
	}
```

**Comment by kidonng** [comment]
@muescha thead, tbody and tbodyRow are all needed. Could you please check the latest code and see if it looks better?

**Comment by muescha** [comment]
to speed up and to have less dom manipulation you can use [`replaceWith`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/replaceWith) and replace it in one step with a new table:

```jsx
let rows = []
...
rows.push(
      <tr>
        {theadCells[i]}
        {tbodyCells[i]}
      </tr>
)
...

table.replaceWith(
	  <table data-table-type="yaml-metadata" className="rgh-vertical-front-matter-table">
		<tbody>{rows}</tbody>
	  </table>
)
```

**Comment by muescha** [comment]
- `thead` is not needed, because in the selector the `table[..] > thead` is included
- `tbody` is not needed, because in the selector the `table[..] > tbody` is included 
- `tbodyRow` is not needed, because in the selector the `table > tbody > tr` is included

you need only the `table` `theadCells` and `tbodyCells`
when you construct in memory and `table.replaceWith` a reference to `tbody` for an `append` is not needed

File: source/features/vertical-front-matter.tsx [hunk 4]
```diff
@@ -19,12 +19,10 @@ function init(): void {
...
 
 	const tbody = table.lastElementChild!;
-	// Table > tbody > tr
-	const tbodyRow = tbody.firstElementChild!;
 	// Table > tbody > tr > td
-	const tbodyCells = [...tbodyRow.children];
+	const tbodyCells = [...tbody.firstElementChild!.children];
 	for (let i = 0; i < theadCells.length; i++) {
-		tbody.append(
+		tbody.prepend(
```
Comment (muescha) [inline source/features/vertical-front-matter.tsx:25]: did this change the order?

**Comment by kidonng** [comment]
Thank you @muescha! Is there anything left?

**Comment by muescha** [comment]
i think this are only milliseconds but with this the css parser has to check **every** tag in the html document for the condition:

```css
[data-table-type="yaml-metadata"]
```

here only check the few `table` tags
```css
table[data-table-type="yaml-metadata"]
```

my count on the example page:
all tags: 985
table tags: 2

**Comment by muescha** [comment]
since you use it on element table:
```js
const tbodyCells = select.all('[data-table-type="yaml-metadata"] > tbody > tr > td', table);
```

then this can be shorter with [`:scope`](https://developer.mozilla.org/en-US/docs/Web/CSS/:scope):
```js
const tbodyCells = select.all(':scope > tbody > tr > td', table);
```

File: source/features/vertical-front-matter.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import * as pageDetect from 'github-url-detection';
+
+import features from '.';
+
+function init(): void {
+	const table = select('[data-table-type="yaml-metadata"]');
+	if (!table) {
+		return;
+	}
+
```
Comment (fregante) [inline source/features/vertical-front-matter.tsx:13]: This can be skipped

Comment (muescha) [reply source/features/vertical-front-matter.tsx:13]: i guess there would be an error when `table` is `undefined` in `table.replaceWith(`

is an early return not always better to save time?

Comment (fregante) [reply source/features/vertical-front-matter.tsx:13]: > i guess there would be an error when `table` is `undefined` in `table.replaceWith(`

Indeed, I thought it was only used here, I didn't see that line

File: source/features/vertical-front-matter.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	const theadCells = select.all('[data-table-type="yaml-metadata"] > thead th', table);
+	if (theadCells.length <= 4) {
+		return;
+	}
+
+	const tbodyCells = select.all('[data-table-type="yaml-metadata"] > tbody > tr > td', table);
```
Comment (fregante) [inline source/features/vertical-front-matter.tsx:19]:
```suggestion
	const tbodyCells = select.all('[data-table-type="yaml-metadata"] > tbody > tr > td');
```

File: source/features/vertical-front-matter.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+import features from '.';
+
+function init(): void {
+	const table = select('[data-table-type="yaml-metadata"]');
+	if (!table) {
+		return;
+	}
+
+	const theadCells = select.all('[data-table-type="yaml-metadata"] > thead th', table);
```
Comment (fregante) [inline source/features/vertical-front-matter.tsx:14]:
```suggestion
	const theadCells = select.all('[data-table-type="yaml-metadata"] > thead th');
```

**Comment by fregante** [comment]
> here only check the few `table` tags(2):

It's not worth following those microoptimizations on a `select` that's run once. The other 2 `select.all` are processed right to left anyway, so the improvement there is also negligible.

**Comment by muescha** [comment]
I just had in mind the issue „make refined faster“ whete the accumulation of  microseconds count

But anyway more readable code is always better

File: source/features/vertical-front-matter.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import './vertical-front-matter.css';
+import React from 'dom-chef';
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+
+import features from '.';
+
+function init(): void {
```
Comment (yakov116) [inline source/features/vertical-front-matter.tsx:8]: Init goes on the bottom of the file

**Comment by kidonng** [comment]
> How about this? Introducing `:scope` doesn't make it particularly shorter

I prefer the `:scope` way 😁

File: source/features/vertical-front-matter.tsx (hunk 8 above)
Comment (kidonng) [reply source/features/vertical-front-matter.tsx:8]: There is only one function (`init`), I don't get you.

Comment (yakov116) [reply source/features/vertical-front-matter.tsx]: Sorry my bad

**Comment by fregante** [comment]
“make refined faster” is specifically about the loading mechanism, which means *acting before dom ready,* which potentially shows features wholes seconds earlier.

**Comment by kidonng** [comment]
Strangely, I didn't get any notifications for that comment 😕

File: source/features/vertical-front-matter.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
+import './vertical-front-matter.css';
+import React from 'dom-chef';
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+
+import features from '.';
+
+// https://github.com/github/markup/blob/cd01f9ec87c86ce5a7c70188a74ef40fc4669c5b/lib/github/markup/markdown.rb#L34
```
Comment (fregante) [inline source/features/vertical-front-matter.tsx:8]: 🙌

**Pull request merged**